apm-lambda-extension/apmproxy/metadata_test.go

```diff
@@ -37,17 +40,106 @@ func Test_processMetadata(t *testing.T) {
 `)
 
-	agentData := apmproxy.AgentData{Data: benchBody, ContentEncoding: ""}
-	extractedMetadata, err := apmproxy.ProcessMetadata(agentData)
-	require.NoError(t, err)
+	testCases := map[string]struct {
+		data         func() []byte
+		encodingType string
+		expectError  error
+	}{
+		"none": {
+			data: func() []byte {
+				return benchBody
+			},
+			encodingType: "",
+		},
+		"gzip": {
+			data: func() []byte {
+				var b bytes.Buffer
+
+				w := gzip.NewWriter(&b)
+
+				_, err := w.Write(benchBody)
+				require.NoError(t, err)
+
+				require.NoError(t, w.Close())
+
+				return b.Bytes()
+			},
+			encodingType: "gzip",
+		},
+		"invalid gzip": {
+			data: func() []byte {
+				return benchBody
+			},
+			expectError:  gzip.ErrHeader,
+			encodingType: "gzip",
+		},
+		"malformed gzip": {
+			data: func() []byte {
+				var b bytes.Buffer
+
+				w := gzip.NewWriter(&b)
+
+				_, err := w.Write(benchBody)
+				require.NoError(t, err)
+
+				return b.Bytes()
+			},
+			expectError:  io.ErrUnexpectedEOF,
+			encodingType: "gzip",
+		},
+		"deflate": {
+			data: func() []byte {
+				var b bytes.Buffer
+
+				w := zlib.NewWriter(&b)
 
-	// Metadata is extracted as is.
-	desiredMetadata := []byte(`{"metadata":{"service":{"name":"1234_service-12a3","version":"5.1.3","environment":"staging","agent":{"name":"elastic-node","version":"3.14.0"},"framework":{"name":"Express","version":"1.2.3"},"language":{"name":"ecmascript","version":"8"},"runtime":{"name":"node","version":"8.0.0"},"node":{"configured_name":"node-123"}},"user":{"username":"bar","id":"123user","email":"bar@user.com"},"labels":{"tag0":null,"tag1":"one","tag2":2},"process":{"pid":1234,"ppid":6789,"title":"node","argv":["node","server.js"]},"system":{"architecture":"x64","hostname":"prod1.example.com","platform":"darwin","container":{"id":"container-id"},"kubernetes":{"namespace":"namespace1","node":{"name":"node-name"},"pod":{"name":"pod-name","uid":"pod-uid"}}},"cloud":{"provider":"cloud_provider","region":"cloud_region","availability_zone":"cloud_availability_zone","instance":{"id":"instance_id","name":"instance_name"},"machine":{"type":"machine_type"},"account":{"id":"account_id","name":"account_name"},"project":{"id":"project_id","name":"project_name"},"service":{"name":"lambda"}}}}`)
-	if err != nil {
-		extension.Log.Errorf("Could not marshal extracted Metadata : %v", err)
+				_, err := w.Write(benchBody)
+				require.NoError(t, err)
+
+				require.NoError(t, w.Close())
+
+				return b.Bytes()
+			},
+			encodingType: "deflate",
+		},
+		"invalid deflate": {
+			data: func() []byte {
+				return benchBody
+			},
+			expectError:  zlib.ErrHeader,
+			encodingType: "deflate",
+		},
+		"malformed deflate": {
```

Comment:
```suggestion
		"incomplete deflate": {
```

Might be a little clearer?

Reply:
Agree 👍